### PR TITLE
LIVY-356 Enabling LDAP authentication for Client to Server.

### PR DESCRIPTION
**Pull request by jakhani**
Currently, Livy doesn't support LDAP Authentication from client(sparkmagic) to server(livy). We need to add LDAP authentication as that's preferable method due to security reasons. We won't be able to use Knox for this purpose. That is why I am raising this PR which contains LDAP authentication.

### CODE REVIEW

File: server/src/main/scala/com/cloudera/livy/server/auth/LdapAuthenticationHandlerImpl.scala [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    env.put("java.naming.security.authentication",
+      LdapAuthenticationHandlerImpl.SECURITY_AUTHENTICATION)
+    env.put("java.naming.security.principal", userDN)
+    env.put("java.naming.security.credentials", password)
+    try {
+      val e = new InitialDirContext(env)
+      e.close()
+      LdapAuthenticationHandlerImpl.logger.debug("Authentication successful for {}", userDN)
+    } catch {
+      case var5: NamingException =>
```
Comment (pranayhasan) [inline server/src/main/scala/com/cloudera/livy/server/auth/LdapAuthenticationHandlerImpl.scala]: What does this var5 mean?

File: server/src/main/scala/com/cloudera/livy/server/auth/LdapAuthenticationHandlerImpl.scala [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+      ctx.addToEnvironment("java.naming.security.principal", userDN)
+      ctx.addToEnvironment("java.naming.security.credentials", password)
+      ctx.lookup(userDN)
+      LdapAuthenticationHandlerImpl.logger.debug("Authentication successful for {}", userDN)
+    } catch {
+      case var13@(_: IOException | _: NamingException) =>
+        throw new AuthenticationException("Error validating LDAP user", var13)
+    } finally if (ctx != null) try ctx.close()
+    catch {
+      case var12: NamingException =>
```
Comment (pranayhasan) [inline server/src/main/scala/com/cloudera/livy/server/auth/LdapAuthenticationHandlerImpl.scala]: Why var12? Does it mean anything?

Comment (jakhani) [inline server/src/main/scala/com/cloudera/livy/server/auth/LdapAuthenticationHandlerImpl.scala]: Corrected. Auto generated name.

File: server/src/main/scala/com/cloudera/livy/server/auth/LdapAuthenticationHandlerImpl.scala (hunk 1 above)
Comment (jakhani) [inline server/src/main/scala/com/cloudera/livy/server/auth/LdapAuthenticationHandlerImpl.scala]: Corrected. Auto generated name.

**Pull request closed without merging**